chainercv/links/model/fpn/head.py

```diff
@@ -0,0 +1,287 @@
+            bbox[:, :, :2] -= bbox[:, :, 2:] / 2
+            bbox[:, :, 2:] += bbox[:, :, :2]
+            # clip
+            bbox[:, :, :2] = self.xp.maximum(bbox[:, :, :2], 0)
+            bbox[:, :, 2:] = self.xp.minimum(
+                bbox[:, :, 2:], self.xp.array(sizes[i]))
+
+            conf = self.xp.exp(conf)
+            score = conf / self.xp.sum(conf, axis=1, keepdims=True)
+
+            bbox, label, score = _suppress(
+                bbox, score, nms_thresh, score_thresh)
+
+            bboxes.append(bbox)
+            labels.append(label)
+            scores.append(score)
+
+        return bboxes, labels, scores
+
+
+class Caffe2FCUniform(chainer.initializer.Initializer):
+
+    def __call__(self, array):
+        scale = 1 / np.sqrt(array.shape[-1])
+        initializers.Uniform(scale)(array)
+
+
+def _suppress(raw_bbox, raw_score, nms_thresh, score_thresh):
+    xp = cuda.get_array_module(raw_bbox, raw_score)
+
+    bbox = []
+    label = []
+    score = []
+    for l in range(raw_score.shape[1] - 1):
+        bbox_l = raw_bbox[:, l + 1]
+        score_l = raw_score[:, l + 1]
+
+        mask = score_l >= score_thresh
+        bbox_l = bbox_l[mask]
+        score_l = score_l[mask]
+
+        order = _argsort(-score_l)
+        bbox_l = bbox_l[order]
+        score_l = score_l[order]
+        indices = utils.non_maximum_suppression(bbox_l, nms_thresh)
+        bbox_l = bbox_l[indices]
+        score_l = score_l[indices]
+
+        bbox.append(bbox_l)
+        label.append(xp.array((l,) * len(bbox_l)))
+        score.append(score_l)
+
+    bbox = xp.vstack(bbox).astype(np.float32)
+    label = xp.hstack(label).astype(np.int32)
+    score = xp.hstack(score).astype(np.float32)
+    return bbox, label, score
+
+
+def head_loss_pre(rois, roi_indices, std, bboxes, labels):
+    thresh = 0.5
+    batchsize_per_image = 512
+    fg_ratio = 0.25
+
+    xp = cuda.get_array_module(*rois)
+
+    n_level = len(rois)
+    roi_levels = xp.hstack(
+        xp.array((l,) * len(rois[l])) for l in range(n_level)).astype(np.int32)
+    rois = xp.vstack(rois).astype(np.float32)
+    roi_indices = xp.hstack(roi_indices).astype(np.int32)
+
+    rois_yx = (rois[:, 2:] + rois[:, :2]) / 2
+    rois_hw = rois[:, 2:] - rois[:, :2]
+    indices = np.unique(cuda.to_cpu(roi_indices))
+
+    gt_locs = xp.empty_like(rois)
+    gt_labels = xp.empty_like(roi_indices)
+    for i in indices:
+        mask = roi_indices == i
+
+        if len(bboxes[i]) > 0:
+            iou = utils.bbox_iou(rois[mask], bboxes[i])
+            gt_index = iou.argmax(axis=1)
+
+            gt_loc = bboxes[i][gt_index].copy()
+        else:
+            gt_loc = xp.empty_like(rois[mask])
+        # tlbr -> yxhw
+        gt_loc[:, 2:] -= gt_loc[:, :2]
+        gt_loc[:, :2] += gt_loc[:, 2:] / 2
+        # offset
+        gt_loc[:, :2] = (gt_loc[:, :2] - rois_yx[mask]) / \
+            rois_hw[mask] / std[0]
+        gt_loc[:, 2:] = xp.log(gt_loc[:, 2:] / rois_hw[mask]) / std[1]
```

Comment:
I find it convenient to see this operation defined in a function.
Could you define one or use the one included already in ChainerCV?
I think your style seems to be simpler than the one below.

https://github.com/chainer/chainercv/blob/master/chainercv/links/model/faster_rcnn/utils/bbox2loc.py